File: scripts/runtime/derive-state.py
#!/usr/bin/env python3
"""Rebuild state.json from events.ndjson for a Circuitry v2 run.

Usage:
    python3 scripts/runtime/derive-state.py <run-root>

Reads circuit.manifest.yaml and events.ndjson, applies the state projection
rules from the v2 spec (Section 5), validates the result against
schemas/state.schema.json, and writes state.json.

Exits 0 on success, 1 on invalid state.
"""

import json
import sys
from datetime import datetime, timezone
from pathlib import Path

import jsonschema
import yaml


REPO_ROOT = Path(__file__).resolve().parents[2]
STATE_SCHEMA_PATH = REPO_ROOT / "schemas" / "state.schema.json"


def load_state_schema() -> dict:
    with open(STATE_SCHEMA_PATH) as f:
        return json.load(f)


def load_manifest(run_root: Path) -> dict:
    manifest_path = run_root / "circuit.manifest.yaml"
    if not manifest_path.exists():
        raise FileNotFoundError(f"circuit.manifest.yaml not found in {run_root}")
    with open(manifest_path) as f:
        return yaml.safe_load(f)


def load_events(run_root: Path) -> list[dict]:
    events_path = run_root / "events.ndjson"
    if not events_path.exists():
        return []
    events = []
    with open(events_path) as f:
        for line in f:
            line = line.strip()
            if line:
                events.append(json.loads(line))
    return events


def derive_state(manifest: dict, events: list[dict]) -> dict:
    """Apply state projection rules over events to produce state.json content.

    This implements the deterministic projection function f(events) -> state
    as specified in Section 5 of the v2 architecture spec.
    """
    circuit = manifest.get("circuit", {})
    circuit_id = circuit.get("id", "")
    manifest_version = circuit.get("version", "")

    # Initialize empty state
    state: dict = {
        "schema_version": "1",
        "run_id": "",
        "circuit_id": circuit_id,
        "manifest_version": manifest_version,
        "status": "initialized",
        "current_step": None,
        "selected_entry_mode": "default",
        "git": {"head_at_start": "0000000"},
        "artifacts": {},
        "jobs": {},
        "checkpoints": {},
    }

    # Internal tracking for step completion (not serialized to state.json)
    # step_completion[step_id] = {"gate_evaluated": bool, "route": str|None}
    step_completion: dict[str, dict] = {}

    for event in events:
        event_type = event.get("event_type", "")
        payload = event.get("payload", {})
        occurred_at = event.get("occurred_at", "")

        # Rule 3: run_started initializes identity, mode, times, git, status
        if event_type == "run_started":
            state["run_id"] = event.get("run_id", "")
            state["selected_entry_mode"] = payload.get("entry_mode", "default")
            state["started_at"] = occurred_at
            state["updated_at"] = occurred_at
            state["git"]["head_at_start"] = payload.get("head_at_start", "0000000")
            state["status"] = "initialized"

        # Rule 4: step_started sets current_step and status
        elif event_type == "step_started":
            step_id = payload.get("step_id", "")
            state["current_step"] = step_id
            state["status"] = "in_progress"
            state["updated_at"] = occurred_at

        # Rule 5: dispatch_requested upserts jobs
        elif event_type == "dispatch_requested":
            step_id = event.get("step_id", payload.get("step_id", ""))
            if not step_id:
                step_id = state.get("current_step", "")
            attempt = payload.get("attempt", 1)
            state["jobs"][step_id] = {
                "attempt": attempt,
                "status": "requested",
                "request": payload.get("request_path", ""),
            }
            state["status"] = "waiting_worker"
            state["updated_at"] = occurred_at

        # Rule 5: dispatch_received upserts jobs
        elif event_type == "dispatch_received":
            step_id = event.get("step_id", payload.get("step_id", ""))
            if not step_id:
                step_id = state.get("current_step", "")
            attempt = payload.get("attempt", 1)
            job = state["jobs"].get(step_id, {"attempt": attempt, "status": "requested"})
            job["status"] = "running"
            job["receipt"] = payload.get("receipt_path", "")
            job["attempt"] = attempt
            state["jobs"][step_id] = job
            state["status"] = "waiting_worker"
            state["updated_at"] = occurred_at

        # Rule 5: job_completed upserts jobs
        elif event_type == "job_completed":
            step_id = event.get("step_id", payload.get("step_id", ""))
            if not step_id:
                step_id = state.get("current_step", "")
            attempt = payload.get("attempt", 1)
            completion = payload.get("completion", "complete")
            job = state["jobs"].get(step_id, {"attempt": attempt, "status": "requested"})
            job["status"] = "complete" if completion == "complete" else "failed"
            job["result"] = payload.get("result_path", "")
            job["attempt"] = attempt
            state["jobs"][step_id] = job
            state["status"] = "in_progress"
            state["updated_at"] = occurred_at

        # Rule 7: checkpoint_requested upserts checkpoints
        elif event_type == "checkpoint_requested":
            step_id = event.get("step_id", payload.get("step_id", ""))
            if not step_id:
                step_id = state.get("current_step", "")
            attempt = payload.get("attempt", 1)
            state["checkpoints"][step_id] = {
                "attempt": attempt,
                "status": "waiting",
                "request_path": payload.get("request_path", ""),
            }
            state["status"] = "waiting_checkpoint"
            state["updated_at"] = occurred_at

        # Rule 7: checkpoint_resolved upserts checkpoints
        elif event_type == "checkpoint_resolved":
            step_id = event.get("step_id", payload.get("step_id", ""))
            if not step_id:
                step_id = state.get("current_step", "")
            attempt = payload.get("attempt", 1)
            cp = state["checkpoints"].get(step_id, {"attempt": attempt, "status": "waiting"})
            cp["status"] = "resolved"
            cp["response_path"] = payload.get("response_path", "")
            cp["selection"] = payload.get("selection", "")
            cp["attempt"] = attempt
            state["checkpoints"][step_id] = cp
            state["status"] = "in_progress"
            state["updated_at"] = occurred_at

        # Rule 8: artifact_written updates artifacts
        elif event_type == "artifact_written":
            artifact_path = payload.get("artifact_path", "")
            step_id = event.get("step_id", "")
            if not step_id:
                step_id = state.get("current_step", "")
            state["artifacts"][artifact_path] = {
                "status": "complete",
                "gate": "pending",
                "produced_by": step_id or "",
                "updated_at": occurred_at,
            }
            state["updated_at"] = occurred_at

        # Rule 8/9: gate_passed updates artifacts and marks step complete
        elif event_type == "gate_passed":
            gate_step_id = payload.get("step_id", "")
            route = payload.get("route", "")
            # Update artifact gate status for artifacts produced by this step
            for art_path, art_info in state["artifacts"].items():
                if art_info.get("produced_by") == gate_step_id:
                    art_info["gate"] = "pass"
                    art_info["updated_at"] = occurred_at
            # Track step completion
            step_completion[gate_step_id] = {"gate_evaluated": True, "route": route}
            state["updated_at"] = occurred_at

        # Rule 8/9: gate_failed updates artifacts
        elif event_type == "gate_failed":
            gate_step_id = payload.get("step_id", "")
            route = payload.get("route", "")
            for art_path, art_info in state["artifacts"].items():
                if art_info.get("produced_by") == gate_step_id:
                    art_info["gate"] = "fail"
                    art_info["updated_at"] = occurred_at
            # A gate_failed with a terminal route still marks the step complete
            step_completion[gate_step_id] = {"gate_evaluated": True, "route": route}
            state["updated_at"] = occurred_at

        # Rule 10: step_reopened resets step, marks artifacts stale
        elif event_type == "step_reopened":
            to_step = payload.get("to_step", "")
            # Reset step completion
            if to_step in step_completion:
                del step_completion[to_step]
            # Mark artifacts produced by the reopened step as stale
            for art_path, art_info in state["artifacts"].items():
                if art_info.get("produced_by") == to_step:
                    art_info["status"] = "stale"
                    art_info["gate"] = "pending"
                    art_info["updated_at"] = occurred_at
            # Reset job/checkpoint records for that step (they no longer make it complete)
            if to_step in state["jobs"]:
                del state["jobs"][to_step]
            if to_step in state["checkpoints"]:
                del state["checkpoints"][to_step]
            state["current_step"] = to_step
            state["status"] = "in_progress"
            state["updated_at"] = occurred_at

        # Rule 11: run_completed sets final state
        elif event_type == "run_completed":
            status = payload.get("status", "completed")
            terminal_target = payload.get("terminal_target", "@complete")
            state["status"] = status
            state["terminal_target"] = terminal_target
            state["current_step"] = None
            state["updated_at"] = occurred_at

    return state


def validate_state(state: dict, schema: dict) -> list[str]:
    validator = jsonschema.Draft202012Validator(schema)
    errors = sorted(validator.iter_errors(state), key=lambda e: list(e.path))
    return [f"{'.'.join(str(p) for p in e.absolute_path)}: {e.message}" for e in errors]


def main() -> int:
    if len(sys.argv) < 2:
        print("Usage: python3 scripts/runtime/derive-state.py <run-root>", file=sys.stderr)
        return 1

    run_root = Path(sys.argv[1]).resolve()
    if not run_root.is_dir():
        print(f"Error: run root does not exist: {run_root}", file=sys.stderr)
        return 1

    try:
        manifest = load_manifest(run_root)
    except FileNotFoundError as e:
        print(f"Error: {e}", file=sys.stderr)
        return 1

    events = load_events(run_root)
    state = derive_state(manifest, events)

    schema = load_state_schema()
    errors = validate_state(state, schema)
    if errors:
        print("State validation errors:", file=sys.stderr)
        for err in errors:
            print(f"  - {err}", file=sys.stderr)
        return 1

    state_path = run_root / "state.json"
    with open(state_path, "w") as f:
        json.dump(state, f, indent=2)
        f.write("\n")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/runtime/engine/src/derive-state.ts
/**
 * Derive run state from the manifest snapshot and append-only event log.
 *
 * Library module -- exports pure functions, no CLI concerns.
 *
 * The core function is `deriveState(manifest, events)` which implements
 * the deterministic projection f(events) -> state.
 */

import { readFileSync, writeFileSync, existsSync } from "node:fs";
import { join } from "node:path";
import { parse as parseYaml } from "yaml";
import { loadJsonSchema, validate } from "./schema.js";

// ── Schema loading ──────────────────────────────────────────────────

/**
 * Load the state JSON-Schema from schemas/state.schema.json.
 */
export function loadStateSchema(): object {
  return loadJsonSchema("schemas/state.schema.json");
}

// ── File I/O ────────────────────────────────────────────────────────

/**
 * Load circuit.manifest.yaml from a run root directory.
 */
export function loadManifest(runRoot: string): object {
  const manifestPath = join(runRoot, "circuit.manifest.yaml");
  if (!existsSync(manifestPath)) {
    throw new Error(`circuit.manifest.yaml not found in ${runRoot}`);
  }
  return parseYaml(readFileSync(manifestPath, "utf-8")) as object;
}

/**
 * Load events.ndjson from a run root directory, returning an array of
 * parsed event objects. Returns [] if the file does not exist.
 */
export function loadEvents(runRoot: string): Record<string, unknown>[] {
  const eventsPath = join(runRoot, "events.ndjson");
  if (!existsSync(eventsPath)) {
    return [];
  }
  const content = readFileSync(eventsPath, "utf-8");
  return content
    .split("\n")
    .filter((line) => line.trim().length > 0)
    .map((line) => JSON.parse(line) as Record<string, unknown>);
}

// ── Core projection ─────────────────────────────────────────────────

/**
 * Apply state projection rules over events to produce state.json content.
 *
 * This implements the deterministic projection function f(events) -> state
 * as specified in Section 5 of the v2 architecture spec.
 */
export function deriveState(
  manifest: Record<string, unknown>,
  events: Record<string, unknown>[],
): Record<string, unknown> {
  const circuit = (manifest.circuit ?? {}) as Record<string, unknown>;
  const circuitId = (circuit.id ?? "") as string;
  const manifestVersion = (circuit.version ?? "") as string;

  // Initialize empty state
  const state: Record<string, unknown> = {
    schema_version: "1",
    run_id: "",
    circuit_id: circuitId,
    manifest_version: manifestVersion,
    status: "initialized",
    current_step: null,
    selected_entry_mode: "default",
    git: { head_at_start: "0000000" },
    artifacts: {} as Record<string, Record<string, unknown>>,
    jobs: {} as Record<string, Record<string, unknown>>,
    checkpoints: {} as Record<string, Record<string, unknown>>,
    routes: {} as Record<string, string>,
  };

  const artifacts = state.artifacts as Record<
    string,
    Record<string, unknown>
  >;
  const jobs = state.jobs as Record<string, Record<string, unknown>>;
  const checkpoints = state.checkpoints as Record<
    string,
    Record<string, unknown>
  >;
  const routes = state.routes as Record<string, string>;

  /** Resolve step_id from event, payload, or current_step. Throws if empty. */
  function resolveStepId(
    event: Record<string, unknown>,
    payload: Record<string, unknown>,
    eventType: string,
  ): string {
    let stepId = (event.step_id ?? payload.step_id ?? "") as string;
    if (!stepId) {
      stepId = (state.current_step ?? "") as string;
    }
    if (!stepId) {
      throw new Error(
        `derive-state: ${eventType} event has no step_id and no current_step is set`,
      );
    }
    return stepId;
  }

  for (const event of events) {
    const eventType = (event.event_type ?? "") as string;
    const payload = (event.payload ?? {}) as Record<string, unknown>;
    const occurredAt = (event.occurred_at ?? "") as string;

    // Rule 3: run_started initializes identity, mode, times, git, status
    if (eventType === "run_started") {
      state.run_id = (event.run_id ?? "") as string;
      state.selected_entry_mode = (payload.entry_mode ?? "default") as string;
      if (typeof payload.goal === "string" && payload.goal.length > 0) {
        state.goal = payload.goal;
      }
      state.started_at = occurredAt;
      state.updated_at = occurredAt;
      (state.git as Record<string, unknown>).head_at_start =
        (payload.head_at_start ?? "0000000") as string;
      state.status = "initialized";
    }

    // Rule 4: step_started sets current_step and status
    else if (eventType === "step_started") {
      const stepId = (payload.step_id ?? "") as string;
      state.current_step = stepId;
      state.status = "in_progress";
      state.updated_at = occurredAt;
    }

    // Rule 5: dispatch_requested upserts jobs
    else if (eventType === "dispatch_requested") {
      const stepId = resolveStepId(event, payload, eventType);
      const attempt = (payload.attempt ?? 1) as number;
      jobs[stepId] = {
        attempt,
        status: "requested",
        request: (payload.request_path ?? "") as string,
      };
      state.status = "waiting_worker";
      state.updated_at = occurredAt;
    }

    // Rule 5: dispatch_received upserts jobs
    else if (eventType === "dispatch_received") {
      const stepId = resolveStepId(event, payload, eventType);
      const attempt = (payload.attempt ?? 1) as number;
      const job: Record<string, unknown> = jobs[stepId] ?? {
        attempt,
        status: "requested",
      };
      job.status = "running";
      job.receipt = (payload.receipt_path ?? "") as string;
      job.attempt = attempt;
      jobs[stepId] = job;
      state.status = "waiting_worker";
      state.updated_at = occurredAt;
    }

    // Rule 5: job_completed upserts jobs
    else if (eventType === "job_completed") {
      const stepId = resolveStepId(event, payload, eventType);
      const attempt = (payload.attempt ?? 1) as number;
      const completion = (payload.completion ?? "complete") as string;
      const job: Record<string, unknown> = jobs[stepId] ?? {
        attempt,
        status: "requested",
      };
      // Preserve exact completion semantics: "complete" -> complete,
      // "partial" -> failed with completion=partial,
      // "blocked" -> failed with completion=blocked
      job.status = completion === "complete" ? "complete" : "failed";
      job.completion = completion;
      if (typeof payload.verdict === "string" && payload.verdict.length > 0) {
        job.verdict = payload.verdict;
      }
      job.result = (payload.result_path ?? "") as string;
      job.attempt = attempt;
      jobs[stepId] = job;
      state.status = "in_progress";
      state.updated_at = occurredAt;
    }

    // Rule 7: checkpoint_requested upserts checkpoints
    else if (eventType === "checkpoint_requested") {
      const stepId = resolveStepId(event, payload, eventType);
      const attempt = (payload.attempt ?? 1) as number;
      checkpoints[stepId] = {
        attempt,
        status: "waiting",
        request_path: (payload.request_path ?? "") as string,
      };
      state.status = "waiting_checkpoint";
      state.updated_at = occurredAt;
    }

    // Rule 7: checkpoint_resolved upserts checkpoints
    else if (eventType === "checkpoint_resolved") {
      const stepId = resolveStepId(event, payload, eventType);
      const attempt = (payload.attempt ?? 1) as number;
      const cp: Record<string, unknown> = checkpoints[stepId] ?? {
        attempt,
        status: "waiting",
      };
      cp.status = "resolved";
      cp.response_path = (payload.response_path ?? "") as string;
      cp.selection = (payload.selection ?? "") as string;
      cp.attempt = attempt;
      checkpoints[stepId] = cp;
      state.status = "in_progress";
      state.updated_at = occurredAt;
    }

    // Rule 8: artifact_written updates artifacts
    else if (eventType === "artifact_written") {
      const artifactPath = (payload.artifact_path ?? "") as string;
      const stepId = resolveStepId(event, payload, eventType);
      artifacts[artifactPath] = {
        status: "complete",
        gate: "pending",
        produced_by: stepId,
        updated_at: occurredAt,
      };
      state.updated_at = occurredAt;
    }

    // Rule 8/9: gate_passed updates artifacts and marks step complete
    else if (eventType === "gate_passed") {
      const gateStepId = (payload.step_id ?? "") as string;
      const route = (payload.route ?? "") as string;
      // Update artifact gate status for artifacts produced by this step
      for (const artInfo of Object.values(artifacts)) {
        if (artInfo.produced_by === gateStepId) {
          artInfo.gate = "pass";
          artInfo.updated_at = occurredAt;
        }
      }
      if (route) {
        routes[gateStepId] = route;
      }
      state.updated_at = occurredAt;
    }

    // Rule 8/9: gate_failed updates artifacts
    else if (eventType === "gate_failed") {
      const gateStepId = (payload.step_id ?? "") as string;
      const route = (payload.route ?? "") as string;
      for (const artInfo of Object.values(artifacts)) {
        if (artInfo.produced_by === gateStepId) {
          artInfo.gate = "fail";
          artInfo.updated_at = occurredAt;
        }
      }
      if (route) {
        routes[gateStepId] = route;
      }
      state.updated_at = occurredAt;
    }

    // Rule 11: run_completed sets final state
    else if (eventType === "run_completed") {
      const status = (payload.status ?? "completed") as string;
      const terminalTarget = (payload.terminal_target ?? "@complete") as string;
      state.status = status;
      state.terminal_target = terminalTarget;
      state.current_step = null;
      state.updated_at = occurredAt;
    }
    else if (eventType === "run_aborted") {
      state.status = "aborted";
      state.current_step = null;
      state.updated_at =
        typeof payload.aborted_at === "string" && payload.aborted_at.length > 0
          ? payload.aborted_at
          : occurredAt;
      if (typeof payload.reason === "string" && payload.reason.length > 0) {
        state.abort_reason = payload.reason;
      }
    }
  }

  return state;
}

// ── Validation ──────────────────────────────────────────────────────

/**
 * Validate a state object against the state JSON-Schema.
 * Returns an array of human-readable error strings (empty = valid).
 */
export function validateState(state: object, schema: object): string[] {
  return validate(schema, state);
}

// ── Output ──────────────────────────────────────────────────────────

export interface DeriveValidatedStateOptions {
  persist?: boolean;
}

export function deriveValidatedStateFromRun(
  runRoot: string,
  options: DeriveValidatedStateOptions = {},
): Record<string, any> {
  const manifest = loadManifest(runRoot) as Record<string, unknown>;
  const events = loadEvents(runRoot);
  const state = deriveState(manifest, events);
  const errors = validateState(state, loadStateSchema());

  if (errors.length > 0) {
    throw new Error(`state validation failed: ${errors.join("; ")}`);
  }

  if (options.persist) {
    writeState(runRoot, state);
  }

  return state as Record<string, any>;
}

/**
 * Write state.json to a run root directory (2-space indent + trailing newline).
 */
export function writeState(runRoot: string, state: object): void {
  const statePath = join(runRoot, "state.json");
  writeFileSync(statePath, JSON.stringify(state, null, 2) + "\n", "utf-8");
}
